plutus-ledger-api/plutus-ledger-api.cabal

```diff
@@ -209,3 +210,22 @@ executable evaluation-test
     , tasty-hunit
 
   default-language: Haskell2010
+
+executable analyse-script-events
+  import:           lang
+  main-is:          Main.hs
+  other-modules:    LoadScriptEvents
+  hs-source-dirs:   exe/analyse-script-events exe/common
+  ghc-options:      -threaded -rtsopts -with-rtsopts=-N
```

Comment:
I'd be very curious to see the output of (if I remember the syntax correctly)

```
  ghc-options:      -threaded -rtsopts -with-rtsopts="-N -s"
```

Particularly if `mapConcurrently` is kept in the code.

Maybe on some smaller inputs in order not to spend a lot of time generating it.

Reply:
> I'd be very curious to see the output of (if I remember the syntax correctly)

Ah, I just copied that without thinking about it.  I'll try your suggestion and see what happens. However `mapConcurrently` is in the script dump executable, which has to be run on a remote machine and takes days to execute if you start from scratch, so it'd be hard to check that.  Marconi may save us from that in future.

Reply:
I ran the `redeemers` analysis (which doesn't have to look inside the scripts) on 53 of the 445 files.  It says
```
1,096,063,320,384 bytes allocated in the heap
 268,349,379,000 bytes copied during GC
   2,469,214,328 bytes maximum residency (145 sample(s))
      78,430,960 bytes maximum slop
            5869 MiB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0     258542 colls,     0 par   203.207s  204.410s     0.0008s    0.0041s
  Gen  1       145 colls,     0 par   96.468s  96.477s     0.6654s    2.2152s

  TASKS: 4 (1 bound, 3 peak workers (3 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.001s  (  0.001s elapsed)
  MUT     time  262.259s  (261.491s elapsed)
  GC      time  299.675s  (300.887s elapsed)
  EXIT    time    0.001s  (  0.001s elapsed)
  Total   time  561.936s  (562.380s elapsed)

  Alloc rate    4,179,318,586 bytes per MUT second

  Productivity  46.7% of total user, 46.5% of total elapsed
```

Reply:
For `values` (which produces a lot more output) it says 
```
1,646,561,997,808 bytes allocated in the heap
 281,399,672,032 bytes copied during GC
   2,489,014,000 bytes maximum residency (125 sample(s))
      93,452,560 bytes maximum slop
            5997 MiB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0     390969 colls,     0 par   217.989s  219.639s     0.0006s    0.0035s
  Gen  1       125 colls,     0 par   99.181s  99.191s     0.7935s    1.9757s

  TASKS: 4 (1 bound, 3 peak workers (3 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.001s  (  0.001s elapsed)
  MUT     time  492.746s  (489.342s elapsed)
  GC      time  317.170s  (318.831s elapsed)
  EXIT    time    0.001s  (  0.007s elapsed)
  Total   time  809.918s  (808.181s elapsed)

  Alloc rate    3,341,601,067 bytes per MUT second

  Productivity  60.8% of total user, 60.5% of total elapsed
```

and for `script-data` (the one that traverses the ASTs) which took about 25 minutes to process 53 files and about 2.6 million scripts) it says

```
4,673,808,869,168 bytes allocated in the heap
 356,110,757,080 bytes copied during GC
   2,428,163,000 bytes maximum residency (114 sample(s))
      84,860,080 bytes maximum slop
            5832 MiB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0     1116718 colls,     0 par   297.789s  300.531s     0.0003s    0.0034s
  Gen  1       114 colls,     0 par   93.369s  93.382s     0.8191s    2.2040s

  TASKS: 4 (1 bound, 3 peak workers (3 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.001s  (  0.001s elapsed)
  MUT     time  1077.909s  (1072.075s elapsed)
  GC      time  391.159s  (393.913s elapsed)
  EXIT    time    0.002s  (  0.001s elapsed)
  Total   time  1469.070s  (1465.990s elapsed)

  Alloc rate    4,335,997,153 bytes per MUT second

  Productivity  73.4% of total user, 73.1% of total elapsed
```

Reply:
Thanks a lot! 

These are not very good numbers, but they're probably not horrible either. Normally more than half the time spent during GC isn't a good thing at all (productivity numbers in the first output), however most of that is spent in Gen 0 which means that either we can't avoid it other than by GCing less often or the GC is triggered too often due to Gen 1 being way too big and eating up all the memory.

I'm not sure how much we care about performance of those executables and if we should make a separate task for optimizing them, but could you please run the following for the first one:

```
  ghc-options:      -threaded -rtsopts -with-rtsopts="-N -A128m -n8m -qb -I0 -s"
```

one last time if it's not too much of a hustle?

Reply:
It turns out that you need to do it like this:
```
  ghc-options:      -O2 -threaded -rtsopts "-with-rtsopts=-N -A128m -n8m -qb -I0 -s"
```
I'm sure there's a good explanation, but it feels like they're doing it on purpose to annoy us.

Anyway, here are the results (this took about 9½ minutes and the machine has 64GB of RAM, if that's relevant).
```
1,093,645,705,664 bytes allocated in the heap
 228,341,430,416 bytes copied during GC
   2,627,887,088 bytes maximum residency (80 sample(s))
      89,431,144 bytes maximum slop
            8107 MiB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0       441 colls,   441 par   208.210s  208.001s     0.4717s    0.8471s
  Gen  1        80 colls,    79 par   96.707s  96.957s     1.2120s    2.4021s

  Parallel GC work balance: 0.00% (serial 0%, perfect 100%)

  TASKS: 34 (1 bound, 33 peak workers (33 total), using -N16)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.043s  (  0.082s elapsed)
  MUT     time  260.053s  (259.396s elapsed)
  GC      time  304.917s  (304.957s elapsed)
  EXIT    time    0.304s  (  0.004s elapsed)
  Total   time  565.317s  (564.440s elapsed)

  Alloc rate    4,205,476,205 bytes per MUT second

  Productivity  46.0% of total user, 46.0% of total elapsed
```